### ISSUE TITLE
libwebsockets+libuv callback returning non-zero values causes sigsegv

### ISSUE THREAD

**Issue by aleqs** [warmcat/libwebsockets#2168]
Hey Andy, I stumbled upon an issue where returning any non-zero value from a callback when LWS_SERVER_OPTION_LIBUV is set, leads to this being printed:

> [2021/01/18 19:28:36:7637] E: internal signal handler caught signal 11
> [2021/01/18 19:28:36:7637] E: lws_libuv_stop

and lws_service() never returns, while if uv is disabled lws_service() happily returns.

It looks like upon returning a non-zero value from the callback, the context gets partially destroyed but uv loop carries on. Replacing any callback branch (e.g. LWS_CALLBACK_CLIENT_WRITEABLE) in any example with "return -1;" leads to this for me. Is this behaviour intended?

**Comment by lws-team** [comment]
Can you show me how to reproduce this on a minimal example on main branch?

**Comment by aleqs** [comment]
Sure, these steps work for me if we modify say lws-minimal-ws-client-binance:
- add to the switch in the callback:
>         case LWS_CALLBACK_CLIENT_WRITEABLE: {
>             return -1;
>         }
> 
- add just before the break in case LWS_CALLBACK_CLIENT_RECEIVE:
>         lws_callback_on_writable(wsi);

This doesn't quite seem to produce sigsegv on main for me, but still it doesn't exit lws_service()

**Comment by aleqs** [comment]
And of course add libuv in:
> info.options = LWS_SERVER_OPTION_DO_SSL_GLOBAL_INIT | LWS_SERVER_OPTION_LIBUV;

**Comment by lws-team** [comment]
There are a few things mixed together here

 - what you describe should just keep dropping the wsi, it should not exit the loop.  Without involving libuv, that's what it does OK.

 - there's a bug in the binance sample that it doesn't cancel the sul when the wsi closes... the sul exists in the wsi's pss and goes out of scope there.  I pushed a patch on main adding it, it's simple enough:

```
diff --git a/minimal-examples/ws-client/minimal-ws-client-binance/main.c b/minimal-examples/ws-client/minimal-ws-client-binance/main.c
index a8ecadf808..2aa31ca19b 100644
--- a/minimal-examples/ws-client/minimal-ws-client-binance/main.c
+++ b/minimal-examples/ws-client/minimal-ws-client-binance/main.c
@@ -296,6 +296,7 @@ callback_minimal(struct lws *wsi, enum lws_callback_reasons reason,
                break;
 
        case LWS_CALLBACK_CLIENT_CLOSED:
+               lws_sul_cancel(&mco->sul_hz);
                goto do_retry;
 
        default:
```

 - there are big changes on main unifying the destroy flow into a common state machine, it seems it's not quite complete for internal event lib loop exit.  I pushed another small patch that makes libuv internal flow also use the unified destroy stuff, that gets it most of the way but you need to do ^C twice to exit and the second one exits without completing the cleanup atm.  But there are no crashes.

So please try current main branch that should solve most of it.

**Comment by aleqs** [comment]
Thank you for getting back to me so quickly. I'll have a look.

Is there a suggested way to break libuv loop from within libwebsockets?

**Comment by aleqs** [comment]
On an unrelated note, my compiler issues an error on main:

> libwebsockets-src/lib/plat/unix/unix-caps.c:44:21: error: implicit conversion changes signedness: 'int' to 'cap_flag_t' [-Werror,-Wsign-conversion]
>         cap_set_flag(caps, mode, count, cv, CAP_SET);

A simple cast fixes it, not sure if it's the right fix though

**Comment by aleqs** [comment]
I think the issue remains, and this time I can formulate the conditions for it more accurately and give you a more specific example.

If I call lws_cancel_service() from another thread and return -1 upon handling the LWS_CALLBACK_EVENT_WAIT_CANCELLED event, the subsequent lws_context_destroy() results in SIGSEGV. I checked the doumentation again and it clearly states that lws_cancel_service() may be called from another thread.

You can reproduce the issue if you replace the original lws-minimal-ws-client-binance example with the code below from the start of the callback until the end of file. I added a lock on context destruction/lws_cancel_service() invocation to eliminate the racing condition where a notification would happen with the context being destroyed because I wasn't sure how this was handled internally.

```
static int
callback_minimal(struct lws *wsi, enum lws_callback_reasons reason,
                 void *user, void *in, size_t len) {
    struct my_conn *mco = (struct my_conn *) user;
    uint64_t latency_us, now_us;
    uint64_t price;
    char numbuf[16];
    const char *p;
    size_t alen;

    switch (reason) {

        case LWS_CALLBACK_CLIENT_CONNECTION_ERROR:
            lwsl_err("CLIENT_CONNECTION_ERROR: %s\n",
                     in ? (char *) in : "(null)");
            goto do_retry;
            break;

        case LWS_CALLBACK_EVENT_WAIT_CANCELLED:
            lwsl_notice("LWS_CALLBACK_EVENT_WAIT_CANCELLED\n");
            return -1;

        case LWS_CALLBACK_CLIENT_RECEIVE:
            // lwsl_hexdump_notice(in, len);

            now_us = (uint64_t) get_us_timeofday();

            p = lws_json_simple_find((const char *) in, len,
                                     "\"depthUpdate\"", &alen);
            if (!p)
                break;

            p = lws_json_simple_find((const char *) in, len, "\"E\":", &alen);
            if (!p) {
                lwsl_err("%s: no E JSON\n", __func__);
                break;
            }
            lws_strnncpy(numbuf, p, alen, sizeof(numbuf));
            latency_us = now_us -
                         ((uint64_t) atoll(numbuf) * LWS_US_PER_MS);

            if (latency_us < mco->e_lat_range.lowest)
                mco->e_lat_range.lowest = latency_us;
            if (latency_us > mco->e_lat_range.highest)
                mco->e_lat_range.highest = latency_us;

            mco->e_lat_range.sum += latency_us;
            mco->e_lat_range.samples++;

            p = lws_json_simple_find((const char *) in, len,
                                     "\"a\":[[\"", &alen);
            if (p) {
                lws_strnncpy(numbuf, p, alen, sizeof(numbuf));
                price = pennies(numbuf);

                if (price < mco->price_range.lowest)
                    mco->price_range.lowest = price;
                if (price > mco->price_range.highest)
                    mco->price_range.highest = price;

                mco->price_range.sum += price;
                mco->price_range.samples++;
            }
            lws_callback_on_writable(wsi);
            break;

        case LWS_CALLBACK_CLIENT_ESTABLISHED:
            lwsl_user("%s: established\n", __func__);
            lws_sul_schedule(lws_get_context(wsi), 0, &mco->sul_hz,
                             sul_hz_cb, LWS_US_PER_SEC);
            mco->wsi = wsi;
            range_reset(&mco->e_lat_range);
            range_reset(&mco->price_range);
            break;

        case LWS_CALLBACK_CLIENT_CLOSED:
            lws_sul_cancel(&mco->sul_hz);
            goto do_retry;

        default:
            break;
    }

    return lws_callback_http_dummy(wsi, reason, user, in, len);

    do_retry:
    if (lws_retry_sul_schedule_retry_wsi(wsi, &mco->sul, connect_client,
                                         &mco->retry_count)) {
        lwsl_err("%s: connection attempts exhausted\n", __func__);
        interrupted = 1;
    }

    return 0;
}

static const struct lws_protocols protocols[] = {
        {"lws-minimal-client", callback_minimal, 0, 0,},
        {NULL, NULL,                             0, 0}
};

static void
sigint_handler(int sig) {
    interrupted = 1;
}

int main(int argc, const char **argv) {
    struct lws_context_creation_info info;
    int n = 0;

    signal(SIGINT, sigint_handler);
    memset(&info, 0, sizeof info);
    lws_cmdline_option_handle_builtin(argc, argv, &info);

    lwsl_user("LWS minimal binance client\n");

    info.options = LWS_SERVER_OPTION_DO_SSL_GLOBAL_INIT;//| LWS_SERVER_OPTION_LIBUV;
    info.port = CONTEXT_PORT_NO_LISTEN;
    info.protocols = protocols;
    info.fd_limit_per_thread = 1 + 1 + 1;
    info.extensions = extensions;

#if defined(LWS_WITH_MBEDTLS) || defined(USE_WOLFSSL)
    info.client_ssl_ca_mem = ca_pem_digicert_global_root;
    info.client_ssl_ca_mem_len = (unsigned int)strlen(ca_pem_digicert_global_root);
#endif

    context = lws_create_context(&info);
    if (!context) {
        lwsl_err("lws init failed\n");
        return 1;
    }

    lws_sul_schedule(context, 0, &mco.sul, connect_client, 1);


    std::mutex context_mutex;
    std::atomic<bool> stop_thread = false;
    auto thread = std::thread([&context_mutex, &stop_thread] {
        while (!stop_thread.load(std::memory_order_acquire)) {
            std::this_thread::sleep_for(std::chrono::seconds(2));
            std::unique_lock<std::mutex> lk(context_mutex);
            if (context != nullptr) {
                lwsl_user("Cancelling service from another thread...\n");
                lws_cancel_service(context);
            }
        }
    });

    std::size_t count = 1000;
    while (n >= 0 && !interrupted && count > 0) {
        n = lws_service(context, 0);
        --count;
    }

    {
        std::unique_lock<std::mutex> lk(context_mutex);
        lws_context_destroy(context);
        context = nullptr;
    }
    stop_thread.store(true, std::memory_order_relaxed);
    thread.join();
    lwsl_user("Completed\n");

    return 0;
}
```

Also, when I enable libuv by adding LWS_SERVER_OPTION_LIBUV in the options and return -1 from the callback, lws_service() never returns. If I understood you correctly, you referred earlier to the fact that it's not necessarily supposed to exit lws_service() when returning -1. Is there a suggested way to to exit lws_service() in the libuv mode?

**Comment by lws-team** [comment]
Yes just wait a bit longer

**Comment by aleqs** [comment]
Sure, waiting...

**Comment by lws-team** [comment]
I updated main with some fixes and cleanup around this.

In addition I pushed the branch _temp, this is main + three patches on top

 - hack on binance example to keep breaking the connection as described in this issue https://libwebsockets.org/git/libwebsockets/commit?id=35ee358f349e3f27ce865d3272c4bf2e80e541b3
 - hack on top of that to convert binance example to libuv foreign loop https://libwebsockets.org/git/libwebsockets/commit?id=b45af4213eade096c6b8a56c5e866d05fe77e392
- optional hack on top of that to convert it more into libuv internal loop https://libwebsockets.org/git/libwebsockets/commit?id=b667b2f4ad1d9cafc685c36cec851c3dc692d93f

These all work well for me, give these scenarios a try.

**Comment by aleqs** [comment]
Let me check...

**Comment by aleqs** [comment]
Thank you for fixing the issues. Unfortunately, it seems like the underlying issue is still there.

On the _temp branch running without libuv and using the code snippet I quoted above (sending the service cancel even from another thread and returning -1 in the callback), I still get sigsegv. Here's the stacktrace:

> #4    Source "main.cc", line 630, in main [0x500e44]
>         628:     {
>         629:         std::unique_lock<std::mutex> lk(context_mutex);
>       > 630:         lws_context_destroy(context);
>         631:         context = nullptr;
>         632:     }
>         633:     stop_thread.store(true, std::memory_order_relaxed);
> #3    Source "libwebsockets-src/lib/core/context.c", line 1804, in lws_context_destroy [0x607467]
>        1801: 		pt = context->pt;
>        1802: 		for (n = 0; n < context->count_threads; n++, pt++) {
>        1803: 			pt->event_loop_pt_unused = 1;
>       >1804: 			lws_pt_destroy(pt);
>        1805: 		}
>        1806: #endif
>        1807: 		/* fallthru */
> #2    Source "libwebsockets-src/lib/core/context.c", line 1554, in lws_pt_destroy [0x607e4d]
>        1552: 	lws_pt_lock(pt, __func__);
>        1553: 	if (pt->pipe_wsi) {
>       >1554: 		lws_destroy_event_pipe(pt->pipe_wsi);
>        1555: 		pt->pipe_wsi = NULL;
>        1556: 	}
> #1    Source "libwebsockets-src/lib/core-net/vhost.c", line 997, in lws_destroy_event_pipe [0x6144be]
>         995: 	lwsl_info("%s\n", __func__);
>         996: 
>       > 997: 	n = lws_wsi_extract_from_loop(wsi);
>         998: 	lws_plat_pipe_close(wsi);
>         999: 	if (!n)
>        1000: 		lws_free(wsi);
> #0    Source "libwebsockets-src/lib/core-net/wsi.c", line 279, in lws_wsi_extract_from_loop [0x61922d]
>         276: 	if (lws_socket_is_valid(wsi->desc.sockfd))
>         277: 		__remove_wsi_socket_from_fds(wsi);
>         278: 
>       > 279: 	if (!wsi->a.context->event_loop_ops->destroy_wsi &&
>         280: 	    wsi->a.context->event_loop_ops->wsi_logical_close) {
>         281: 		wsi->a.context->event_loop_ops->wsi_logical_close(wsi);
>         282: 		return 1; /* close / destroy continues async */
> Segmentation fault (Address not mapped to object [0x720])
> 

I'm yet to try the libuv loop example you kindly added

**Comment by lws-team** [comment]
Yes the way is check the things I provided that work well as far as I can tell.  If the work well for you too, align your code to them.  If not, let's discuss the problem on the code we both have.

**Comment by aleqs** [comment]
Migrating the code...

**Comment by aleqs** [comment]
The updated example with libuv runs well as is apart from the fact that it needs a signal to be stopped. Is there a way to get libuv to break the loop?

When I modify the code in line with what I posted above:
- calling lws_cancel_service() from another thread every 2sec
- printing and returning -1 on LWS_CALLBACK_EVENT_WAIT_CANCELLED
- breaking on LWS_CALLBACK_CLIENT_WRITEABLE instead of returning -1

I get a peculiar behavour where LWS_CALLBACK_EVENT_WAIT_CANCELLED gets triggered only once.


This is the callback:
```
static int
callback_minimal(struct lws *wsi, enum lws_callback_reasons reason,
                 void *user, void *in, size_t len)
{
    struct my_conn *mco = (struct my_conn *)user;
    uint64_t latency_us, now_us;
    uint64_t price;
    char numbuf[16];
    const char *p;
    size_t alen;

    switch (reason) {

        case LWS_CALLBACK_CLIENT_CONNECTION_ERROR:
            lwsl_err("CLIENT_CONNECTION_ERROR: %s\n",
                     in ? (char *)in : "(null)");
            goto do_retry;
            break;

        case LWS_CALLBACK_CLIENT_RECEIVE:
            /*
             * The messages are a few 100 bytes of JSON each
             */

            // lwsl_hexdump_notice(in, len);

            now_us = (uint64_t)get_us_timeofday();

            p = lws_json_simple_find((const char *)in, len,
                                     "\"depthUpdate\"", &alen);
            /*
             * Only the JSON with depthUpdate init has the numbers we care
             * about as well
             */
            if (!p)
                break;

            p = lws_json_simple_find((const char *)in, len, "\"E\":", &alen);
            if (!p) {
                lwsl_err("%s: no E JSON\n", __func__);
                break;
            }
            lws_strnncpy(numbuf, p, alen, sizeof(numbuf));
            latency_us = now_us -
                         ((uint64_t)atoll(numbuf) * LWS_US_PER_MS);

            if (latency_us < mco->e_lat_range.lowest)
                mco->e_lat_range.lowest = latency_us;
            if (latency_us > mco->e_lat_range.highest)
                mco->e_lat_range.highest = latency_us;

            mco->e_lat_range.sum += latency_us;
            mco->e_lat_range.samples++;

            p = lws_json_simple_find((const char *)in, len,
                                     "\"a\":[[\"", &alen);
            if (p) {
                lws_strnncpy(numbuf, p, alen, sizeof(numbuf));
                price = pennies(numbuf);

                if (price < mco->price_range.lowest)
                    mco->price_range.lowest = price;
                if (price > mco->price_range.highest)
                    mco->price_range.highest = price;

                mco->price_range.sum += price;
                mco->price_range.samples++;
            }

            lws_callback_on_writable(wsi);
            break;

        case LWS_CALLBACK_CLIENT_ESTABLISHED:
            lwsl_user("%s: established\n", __func__);
            lws_sul_schedule(lws_get_context(wsi), 0, &mco->sul_hz,
                             sul_hz_cb, LWS_US_PER_SEC);
            mco->wsi = wsi;
            range_reset(&mco->e_lat_range);
            range_reset(&mco->price_range);
            break;

        case LWS_CALLBACK_CLIENT_WRITEABLE: {
            //lwsl_notice("LWS_CALLBACK_CLIENT_WRITEABLE\n");
            //return -1;
            break;
        }
        case LWS_CALLBACK_EVENT_WAIT_CANCELLED:
            lwsl_notice("LWS_CALLBACK_EVENT_WAIT_CANCELLED\n");
            return -1;
        case LWS_CALLBACK_CLIENT_CLOSED:
            lws_sul_cancel(&mco->sul_hz);
            goto do_retry;

        default:
            break;
    }

    return lws_callback_http_dummy(wsi, reason, user, in, len);

    do_retry:
    /*
     * retry the connection to keep it nailed up
     *
     * For this example, we try to conceal any problem for one set of
     * backoff retries and then exit the app.
     *
     * If you set retry.conceal_count to be larger than the number of
     * elements in the backoff table, it will never give up and keep
     * retrying at the last backoff delay plus the random jitter amount.
     */
    if (lws_retry_sul_schedule_retry_wsi(wsi, &mco->sul, connect_client,
                                         &mco->retry_count)) {
        lwsl_err("%s: connection attempts exhausted\n", __func__);
        interrupted = 1;
    }

    return 0;
}
```

This is the main:
```
int main(int argc, const char **argv)
{
    struct lws_context_creation_info info;
//	void *foreign_loops[1];

    memset(&info, 0, sizeof info);
    lws_cmdline_option_handle_builtin(argc, argv, &info);

    lwsl_user("LWS minimal binance client\n");

#if 0
    uv_loop_init(&loop_uv);
        uv_signal_init(&loop_uv, &sighandler_uv);
        uv_signal_start(&sighandler_uv, signal_cb_uv, SIGINT);

        foreign_loops[0] = &loop_uv;
        info.foreign_loops = foreign_loops;
#endif

    info.options = LWS_SERVER_OPTION_DO_SSL_GLOBAL_INIT | LWS_SERVER_OPTION_LIBUV;
    info.port = CONTEXT_PORT_NO_LISTEN; /* we do not run any server */
    info.protocols = protocols;
    info.fd_limit_per_thread = 1 + 1 + 1;
    info.extensions = extensions;

#if defined(LWS_WITH_MBEDTLS) || defined(USE_WOLFSSL)
    /*
         * OpenSSL uses the system trust store.  mbedTLS / WolfSSL have to be
         * told which CA to trust explicitly.
         */
        info.client_ssl_ca_mem = ca_pem_digicert_global_root;
        info.client_ssl_ca_mem_len = (unsigned int)strlen(ca_pem_digicert_global_root);
#endif

    context = lws_create_context(&info);
    if (!context) {
        lwsl_err("lws init failed\n");
        return 1;
    }

    std::mutex context_mutex;
    std::atomic<bool> stop_thread = false;
    auto thread = std::thread([&context_mutex, &stop_thread] {
        while (!stop_thread.load(std::memory_order_acquire)) {
            std::this_thread::sleep_for(std::chrono::seconds(2));
            std::unique_lock<std::mutex> lk(context_mutex);
            if (context != nullptr) {
                lwsl_user("Cancelling service from another thread...\n");
                lws_cancel_service(context);
            }
        }
    });

    /* schedule the first client connection attempt to happen immediately */
    lws_sul_schedule(context, 0, &mco.sul, connect_client, 1);

//	uv_run(&loop_uv, UV_RUN_DEFAULT);

    //while (
    lws_service(context, 10000);
    //>= 0)
    //	;

    {
        std::unique_lock<std::mutex> lk(context_mutex);
        lws_context_destroy(context);
        context = nullptr;
    }
    stop_thread.store(true, std::memory_order_relaxed);
    thread.join();
#if 0
    uv_timer_stop(&timer_outer_uv);
        uv_close((uv_handle_t*)&timer_outer_uv, NULL);
        uv_signal_stop(&sighandler_uv);
        uv_close((uv_handle_t *)&sighandler_uv, NULL);

        while (uv_loop_close(&loop_uv) == UV_EBUSY)
                uv_run(&loop_uv, UV_RUN_NOWAIT);
#endif
    lwsl_user("Completed\n");

    return 0;
}
```

**Comment by aleqs** [comment]
> Thank you for fixing the issues. Unfortunately, it seems like the underlying issue is still there.
> 
> On the _temp branch running without libuv and using the code snippet I quoted above (sending the service cancel event from another thread and returning -1 in the callback when the reason is LWS_CALLBACK_EVENT_WAIT_CANCELLED), I still get sigsegv. Here's the stacktrace:
> 
> > #4    Source "main.cc", line 630, in main [0x500e44]
> > 628:     {
> > 629:         std::unique_lockstd::mutex lk(context_mutex);
> > > 630:         lws_context_destroy(context);
> > 631:         context = nullptr;
> > 632:     }
> > 633:     stop_thread.store(true, std::memory_order_relaxed);
> > #3    Source "libwebsockets-src/lib/core/context.c", line 1804, in lws_context_destroy [0x607467]
> > 1801: 		pt = context->pt;
> > 1802: 		for (n = 0; n < context->count_threads; n++, pt++) {
> > 1803: 			pt->event_loop_pt_unused = 1;
> > >1804: 			lws_pt_destroy(pt);
> > 1805: 		}
> > 1806: #endif
> > 1807: 		/* fallthru _/
> > #2    Source "libwebsockets-src/lib/core/context.c", line 1554, in lws_pt_destroy [0x607e4d]
> > 1552: 	lws_pt_lock(pt, **func**);
> > 1553: 	if (pt->pipe_wsi) {
> > >1554: 		lws_destroy_event_pipe(pt->pipe_wsi);
> > 1555: 		pt->pipe_wsi = NULL;
> > 1556: 	}
> > #1    Source "libwebsockets-src/lib/core-net/vhost.c", line 997, in lws_destroy_event_pipe [0x6144be]
> > 995: 	lwsl_info("%s\n", **func**);
> > 996:
> > > 997: 	n = lws_wsi_extract_from_loop(wsi);
> > 998: 	lws_plat_pipe_close(wsi);
> > 999: 	if (!n)
> > 1000: 		lws_free(wsi);
> > #0    Source "libwebsockets-src/lib/core-net/wsi.c", line 279, in lws_wsi_extract_from_loop [0x61922d]
> > 276: 	if (lws_socket_is_valid(wsi->desc.sockfd))
> > 277: 		__remove_wsi_socket_from_fds(wsi);
> > 278:
> > > 279: 	if (!wsi->a.context->event_loop_ops->destroy_wsi &&
> > 280: 	    wsi->a.context->event_loop_ops->wsi_logical_close) {
> > 281: 		wsi->a.context->event_loop_ops->wsi_logical_close(wsi);
> > 282: 		return 1; /_ close / destroy continues async */
> > Segmentation fault (Address not mapped to object [0x720])
> 
> I'm yet to try the libuv loop example you kindly added

Btw this can still be reproduced by removing "| LWS_SERVER_OPTION_LIBUV" with the latest modification to your code I posted. I think this bug is in direct breach of the API. Maybe it could be easier to start the investigation with this one, then you wouldn't need to deal with the libuv code?

**Comment by lws-team** [comment]
The examples have to pick what behaviours they want to showcase, if they try to show all possibilities in one place they just become impenetrable messes.

At least for foreign loop case the canonical example for that uses a native libuv timer to call lws_destroy_context() after 10s of serving to start the process of the context cleaning itself off the foreign loop and destroying itself (while leaving the loop intact).

https://libwebsockets.org/git/libwebsockets/tree/minimal-examples/http-server/minimal-http-server-eventlib-foreign/minimal-http-server-eventlib-foreign.c#n116

```
case LWS_CALLBACK_EVENT_WAIT_CANCELLED:
            lwsl_notice("LWS_CALLBACK_EVENT_WAIT_CANCELLED\n");
            return -1;
```

Returning -1 just hangs up the wsi, and your code schedules a retry, I doubt it's what you want.  If you want to finish with the whole context you should be able to trigger it by replacing return -1 with lws_destroy_context() and wait for lws_cancel_service().

I really recommend having your own explicit foreign loop instead of the internal one.

**Comment by aleqs** [comment]
Thank you - I didn't think I could invoke lws_context_destroy() from within the callback. For both the default and libuv options, it works the way you described and without issues for me on _temp.

Thank you, I'll give foreign loops a go.
